### PR TITLE
[tmf] add Backbone TMF agent

### PR DESCRIPTION
**Pull request by simonlingoogle**
This PR adds the Backbone TMF agent

### CODE REVIEW

**Comment by simonlingoogle** [comment]
Labeld as `P2` because many fore-coming 1.2 Backbone features including MLR, DUA, CCM requires this backbone TMF feature. @wgtdkp @librasungirl @bukepo @jwhui @abtink

File: src/core/thread/tmf.hpp [hunk 1]
```diff
@@ -57,8 +57,9 @@ class TmfAgent : public Coap::Coap
      * @param[in] aInstance      A reference to the OpenThread instance.
      *
      */
-    explicit TmfAgent(Instance &aInstance)
+    explicit TmfAgent(Instance &aInstance, otNetifIdentifier aNetifIdentifier)
```
Comment (bukepo) [inline src/core/thread/tmf.hpp:60]: Is it possible to let this class inherit CoapBase and managing two sockets, one for Thread interface and one for Backbone link interface. I guess this will save some resources.

Comment (simonlingoogle) [reply src/core/thread/tmf.hpp:60]: It will save some RAM resource but the memory saved should be trivial for backbone routers which tends to have megabytes of RAM. 
With one COAP handling two sockets, we are going to make the COAP logic more complex with increased code size (e.g. choosing UDP sockets for sending/receiving messages based on message attributes)
Also with the security issue in mind, we will have to determine whether a resource should handle Backbone requests. 
I would prefer to use separate COAP objects since it's simpler. 
Thoughts?

File: src/core/thread/tmf.hpp [hunk 2]
```diff
@@ -82,8 +83,10 @@ class TmfAgent : public Coap::Coap
     bool IsTmfMessage(const Ip6::MessageInfo &aMessageInfo) const;
 
 private:
-    otError        Start(uint16_t aPort);
+    otError        Start(uint16_t aPort, otNetifIdentifier aNetifIdentifier);
     static otError Filter(const ot::Coap::Message &aMessage, const Ip6::MessageInfo &aMessageInfo, void *aContext);
+
+    otNetifIdentifier mNetifIdentifier;
```
[the remark this answers is not in the document]
Comment (simonlingoogle) [reply src/core/thread/tmf.hpp:89]: Yes it can. I don't have strong opinion on this and there is no significant difference. 

If it's a parameter of `Start()`, we need to change `GetInstance().GetBackboneTmfAgent().Start()` to `GetInstance().GetBackboneTmfAgent().Start(OT_NETIF_BACKBONE)`. It seems a little redundant because `GetInstance().GetBackboneTmfAgent().Start(OT_NETIF_THREAD)` is illegal.

File: src/core/net/udp6.hpp [hunk 3]
```diff
@@ -123,6 +123,10 @@ class Udp : public InstanceLocator
         {
             mHandler(mContext, &aMessage, &aMessageInfo);
         }
+
+#if OPENTHREAD_FTD && OPENTHREAD_CONFIG_BACKBONE_ROUTER_ENABLE
+        bool mBoundToThreadStack;
```
[the remark this answers is not in the document]
Comment (simonlingoogle) [reply src/core/net/udp6.hpp:128]: Yes. Can't pass checks without this. 

The problem was on Backbone Routers. Since a Backbone Router has two Udp Sockets both bound to port 61631, it does not know which one is the correct receiver for a received UDP message. This flag thus help Backbone Router to skip the Udp sockets bound to Backbone interface that are not supposed to process any UDP messages received from the Thread stack.

File: src/core/thread/tmf.hpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (bukepo) [reply src/core/thread/tmf.hpp:89]: I think we should have separate class for BackboneTmfAgent, which not only has a different port and netif, but also different filter rules. Thoughts?

File: src/core/net/udp6.hpp (hunk 3 above)
[the remark this answers is not in the document]
Comment (bukepo) [reply src/core/net/udp6.hpp:128]: This adds the overhead to all sockets. As `MessageInfo::mIstHostInterface` tells whether this message comes from Thread network interface, is it possible to add special handling for the backbone tmf port here?

[the remark this answers is not in the document]
Comment (simonlingoogle) [reply src/core/net/udp6.hpp:128]: This adds the overhead to all sockets, **but only on Backbone Routers**. Note that Backbone Routers are usually way more powerful than normal SOC devices. 

The problem is in the UDP socket matching where UDP sockets know nothing about up layer COAP objects. It does not know which Udp Socket should receive a UDP message if they share the same port. We can add special handling to skip the TMF socket by comparing Socket object references. 

However I think we should not keep adding hardcoding and special handling for Netif binding issues, but to accept that OT core should be aware of different network interfaces (at least on Backbone Routers).

File: src/core/thread/tmf.hpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (simonlingoogle) [reply src/core/thread/tmf.hpp:89]: Note that the default port for Backbone TMF is 61631, same as the Thread TMF port. So, by default they share the same port number. 

I haven't investigated the filter rules. If they differ for Thread TMF and Backbone TMF, I would agree that it's better to have separate classes. 
But I think it's not the major issue. I will investigate and revise accordingly.

File: src/core/net/udp6.hpp (hunk 3 above)
[the remark this answers is not in the document]
Comment (bukepo) [reply src/core/net/udp6.hpp:128]: Since this field is only added into the internal class, I'm wondering if this will affect upper layer users who don't have such internal classes from C API. `otUdpSocket` is directly used in the CLI library, will it cause issues when being cast to the internal `SocketHandle`?

File: src/core/thread/tmf.hpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (bukepo) [reply src/core/thread/tmf.hpp:89]: I guess the backbone TMF should only allow link local scope destinations?

File: src/core/net/udp6.hpp (hunk 3 above)
[the remark this answers is not in the document]
Comment (simonlingoogle) [reply src/core/net/udp6.hpp:128]: This is a good point. Adding `bool mBoundToThreadStack` to `otUdpSocket` brings unnecessary cost that should be avoided.

[the remark this answers is not in the document]
Comment (bukepo) [reply src/core/net/udp6.hpp:128]: How about managing sockets segment by segment.

```bash

socket1 -> socket2 - > ... -> socket[n] -> socket[n+1] -> ...
     ^                            ^
     |                            |
Head of all sockets(not bound)  First backbone link socket
```

File: src/core/common/linked_list.hpp [hunk 4]
```diff
@@ -409,13 +445,17 @@ template <typename Type> class LinkedList
      * @returns A pointer to the matching entry if one is found, or nullptr if no matching entry was found.
      *
      */
-    template <typename Inidcator> const Type *FindMatching(const Inidcator &aIndicator, const Type *&aPrevEntry) const
+    template <typename Inidcator>
+    const Type *FindMatching(const Type *     aBegin,
+                             const Type *     aEnd,
```
Comment (bukepo) [inline src/core/common/linked_list.hpp:450]: Avoid touching unrelated code by making `aEnd` optional?

File: src/posix/platform/udp.cpp [hunk 5]
```diff
@@ -276,9 +276,6 @@ otError otPlatUdpBind(otUdpSocket *aUdpSocket)
         VerifyOrExit(0 == setsockopt(fd, IPPROTO_IPV6, IPV6_RECVPKTINFO, &on, sizeof(on)), error = OT_ERROR_FAILED);
     }
 
-    VerifyOrExit(0 == setsockopt(fd, IPPROTO_IPV6, IPV6_MULTICAST_IF, &gNetifIndex, sizeof(gNetifIndex)),
```
Comment (simonlingoogle) [inline src/posix/platform/udp.cpp:279]: Some explanation for removing this line of code:

- For sockets that bound to Thread network interface, this line of code is useless since the multicast network interface is already fixed to Thread interface only. 
- For sockets that bound to Backbone network interface, this line of code is wrong (will always fail) and thus should be removed. 
- For sockets that bound to unspecified (all) interfaces (e.g. mApplicationCoap), it could be a security pitfall to bind `MULTICAST_IF` to Thread interface because we might accidentally send multicast in the Thread network. I prefer to add `MUTLICAST_IF` configuration when we really need it.

File: src/core/net/udp6.cpp [hunk 6]
```diff
@@ -205,6 +221,35 @@ otError Udp::Bind(SocketHandle &aSocket, const SockAddr &aSockAddr)
...
+
+    if (mPrevBackboneSockets != nullptr)
+    {
+        mSockets.PushAfter(aSocket, *mPrevBackboneSockets);
+    }
+    else
+    {
+        IgnoreError(mSockets.Add(aSocket));
+    }
+}
```
Comment (abtink) [inline src/core/net/udp6.cpp:245]: I think it may be actually much easier and a lot cleaner to use two linked-lists, one as before and one for `BackboneSocket`. This makes all the logic in the core for this much simpler.

I think you mentioned the only limitation is with public `otUdpGetSockets()` behavior.

How about adding a new public API that gets the list of backbone sockets in addition to `otUdpGetSockets()` which would not include the backbone sockets.

We can also later change the public API to an iterator based API to be more flexible.

File: src/core/thread/tmf.hpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (simonlingoogle) [reply src/core/thread/tmf.hpp]: this is resolved since there is now a separate class `BackboneTmfAgent` with customized TMF filter rules.

File: src/core/net/udp6.hpp (hunk 3 above)
[the remark this answers is not in the document]
Comment (simonlingoogle) [reply src/core/net/udp6.hpp]: segmented linked list implemented. Please check.

File: src/core/common/linked_list.hpp [hunk 7]
```diff
@@ -409,13 +453,17 @@ template <typename Type> class LinkedList
...
+                             const Inidcator &aIndicator,
+                             const Type *&    aPrevEntry) const
```
[the remark this answers is not in the document]
Comment (simonlingoogle) [reply src/core/common/linked_list.hpp:460]: Thanks for the two approaches, both of which should work. 
We have thought about `Approach 1` but decided to add a ranged `FindMatching` method to `LinkedList`:
- It's a valid use case to search in a given range of linked list. The method might also be useful for other purposes. 
- It does not increase any code size  if ranged `FindMatching` is not used. (from the size-report we can find that the overall code size actually decreased by 16B)

For `approach 2`, I think it will increase code size so it does not seem to be the best choice. 

Thoughts?

File: src/core/common/linked_list.hpp [hunk 8]
```diff
@@ -348,33 +348,74 @@ template <typename Type> class LinkedList
     /**
      * This method searches within the linked list to find an entry and if found returns a pointer to previous entry.
      *
-     * @param[in]  aEntry      A reference to an entry to find.
+     * @param[in]  aEntry      A pointer to an entry to find. If nullptr, this method finds the tail of the list.
```
[the remark this answers is not in the document]
Comment (simonlingoogle) [reply src/core/common/linked_list.hpp:351]: You are right. 
I was too focused on recalculating `mPrefBackboneSockets` using `mSockets.Find(backboneSockets, mPrevBackboneSockets))`.

File: src/core/net/udp6.cpp (hunk 6 above)
Comment (simonlingoogle) [reply src/core/net/udp6.cpp:245]: The segmented linked list just works. It's largest advantage is that it does not affect the public API at all.
I would prefer to switch to use multiple linked lists of sockets when it's really required.

Thoughts? @abtink @bukepo

Comment (bukepo) [reply src/core/net/udp6.cpp:245]: It seems we are using the segmented list for managing addresses. I like its simplicity for implementation and also allows providing simpler public APIs(introduce another iterator).

Comment (abtink) [reply src/core/net/udp6.cpp:245]: Using segmented lists sounds fine to me.

File: src/core/common/linked_list.hpp (hunk 7 above)
[the remark this answers is not in the document]
Comment (abtink) [reply src/core/common/linked_list.hpp:419]: SGTM. Thanks.

File: src/core/net/udp6.cpp [hunk 9]
```diff
@@ -205,6 +221,35 @@ otError Udp::Bind(SocketHandle &aSocket, const SockAddr &aSockAddr)
...
+{
+    RemoveSocket(aSocket);
+
+    if (mPrevBackboneSockets != nullptr)
+    {
+        mSockets.PushAfter(aSocket, *mPrevBackboneSockets);
+    }
+    else
+    {
+        IgnoreError(mSockets.Add(aSocket));
```
Comment (abtink) [inline src/core/net/udp6.cpp:243]: Use  `mSocket.Push(aSocket)` here?

`Add()` checks for it not to be on the list already before adding/pushing it to the list but since we know it's not in the list (we did `RemoveSocke()` already) we can use `Push` (which does not check).

File: src/core/net/udp6.cpp [hunk 10]
```diff
@@ -297,12 +342,40 @@ otError Udp::SendTo(SocketHandle &aSocket, Message &aMessage, const MessageInfo
...
+
+#if OPENTHREAD_FTD && OPENTHREAD_CONFIG_BACKBONE_ROUTER_ENABLE
+    if (&aSocket == mPrevBackboneSockets)
+    {
+        SocketHandle *backboneSockets = aSocket.GetNext();
+        otError       error;
+
+        if (backboneSockets != nullptr)
+        {
+            IgnoreError(error = mSockets.Find(*backboneSockets, mPrevBackboneSockets));
```
Comment (abtink) [inline src/core/net/udp6.cpp:369]: Remove `IgnoreError()`?

File: src/core/net/udp6.cpp [hunk 11]
```diff
@@ -297,12 +342,40 @@ otError Udp::SendTo(SocketHandle &aSocket, Message &aMessage, const MessageInfo
...
+        mPrevBackboneSockets = &aSocket;
+    }
+#endif
+exit:
+    return;
 }
 
 void Udp::RemoveSocket(SocketHandle &aSocket)
 {
     SuccessOrExit(mSockets.Remove(aSocket));
```
Comment (abtink) [inline src/core/net/udp6.cpp:359]: I think we can do a small optimization here.
 
If instead of `Remove()` we first search for entry in list (using `Find()`) we get the `prev` entry before it, we can then remove it using `PopAfter()` and use it to see update `mPrevBackboneSockets` (if needed).
```
     SocketHandle *prev;

     SuccessOrExit(mSockets.Find(aSocket , prev));
     mSocket.PopAfter(prev);
     
 #if OPENTHREAD_FTD && OPENTHREAD_CONFIG_BACKBONE_ROUTER_ENABLE
     if (&aSocket == mPrevBackboneSockets)
     {
           mPrevBackboneSockets = prev;
     }
#endif 
```

File: src/core/thread/tmf.hpp (hunk 1 above)
Comment (simonlingoogle) [reply src/core/thread/tmf.hpp]: I think we can refactor in the future if `one COAP base two sockets` is proven to be a better idea.

File: src/core/net/udp6.cpp (hunk 10 above)
Comment (simonlingoogle) [reply src/core/net/udp6.cpp:369]: If we remove `IgnoreError()` , we need to add `OT_UNUSED_VARIABLE(error)` (note that `OT_ASSERT` can be turned off). 
I am fine with both, but `IgnoreError()` reduced one line of code.

File: src/core/net/udp6.cpp (hunk 11 above)
Comment (simonlingoogle) [reply src/core/net/udp6.cpp:359]: I didn't put much thought into it. The code you suggested is indeed more efficient. Thanks.

File: src/posix/platform/udp.cpp (hunk 5 above)
Comment (bukepo) [reply src/posix/platform/udp.cpp:279]: I'm fine get back to this issue later.

File: src/core/net/udp6.cpp (hunk 9 above)
Comment (simonlingoogle) [reply src/core/net/udp6.cpp]: revised

File: src/core/net/udp6.cpp (hunk 10 above)
Comment (simonlingoogle) [reply src/core/net/udp6.cpp]: **Just for discussion:** 
If we have a method like 
```c++
static inline void OT_ASSERT_OK(otError aError) 
{
    OT_UNUSED_VARIABLE(aError);
    OT_ASSERT(aError == OT_ERROR_NONE);
}
```
 , we can avoid using `IgnoreError` or `OT_UNUSED_VARIABLE`. It's quite easy (and a little annoying) to miss these statements and get build failures.

**Comment by simonlingoogle** [comment]
**Beyond the scope of this PR:**

I found that allowing `Find()` of linked list to get the tail with `nullptr` seems even reduce the code size. 
**I tend to think that using `nullptr` to get the tail make sense since the next of tail is indeed `nullptr`.**
Then we can implement `GetTail()` by using `Find()` directly, which should further reduce code size. 

Thought? @abtink

**Comment by abtink** [comment]
> Beyond the scope of this PR:
> 
> I found that allowing Find() of linked list to get the tail with nullptr seems even reduce the code size.
> I tend to think that using nullptr to get the tail make sense since the next of tail is indeed nullptr.
> Then we can implement GetTail() by using Find() directly, which should further reduce code size.
> 
> Thought? @abtink 

Sounds reasonable to me. For a future PR. 
I would suggest that we define the version of `Find()` which uses a pointer as input as `private` and use it from different places (e.g. from `GetTail()` and the public `Find()`). This way the public methods are clearly defined (not complex) but common code is shared privately.

File: src/core/net/udp6.cpp (hunk 10 above)
Comment (abtink) [reply src/core/net/udp6.cpp]: I don' think that would work on all toolchains (the `OT_UNUSED_VARIABLE` can apply to local parameter variable `aError` and not the caller one). 

The trickiness in general is that `OT_UNUSED_VARIABLE` is tool-chain dependent.

 I would recommend using `OT_UNUSED_VARIABLE` directly to silence the unused-var warning from such `OT_ASEERT()` use and not use `IgnoreError` for this. `IgnoreError` is to ensure returned value is not ignored and it just happens to work in this case to use the variable (silence the warning), but may not work in all situations and toolchains.

Comment (simonlingoogle) [reply src/core/net/udp6.cpp]: **Note that the code we are commenting is already removed, so resolving this.**

File: src/core/backbone_router/backbone_tmf.cpp [hunk 12]
```diff
@@ -0,0 +1,75 @@
...
+    const Ip6::Address &dst = aMessageInfo.GetSockAddr();
+    const Ip6::Address &src = aMessageInfo.GetPeerAddr();
+
+    // A Backbone TMF message must comply with following rules:
+    // The destination must be one of:
+    //     1. All Network BBRs (Link-Local scope)
+    //     2. All Domain BBRs (Link-Local scope)
+    //     3. A Backbone Link-Local address
+    // The source must be a Backbone Link-local address.
+    return (dst.IsLinkLocal() || dst.IsLinkLocalMulticast()) && src.IsLinkLocal();
```
[the remark this answers is not in the document]
Comment (simonlingoogle) [reply src/core/backbone_router/backbone_tmf.cpp:69]: I was thinking about this too, but thought it might not matter that much. 
 
Note that if the code here runs, the destination of the message **MUST** be a subscribed Multicast Address on the Backbone interface. The other possible Multicast Addresses are `Solicited-node Multicast Addresses` which are also link-local. 
We should never receive TMF messages with `Solicited-node Multicast Addresses` as destinations. But even if we did receive a message like that, it does not harm to process it? 

Thoughts?

[the remark this answers is not in the document]
Comment (librasungirl) [reply src/core/backbone_router/backbone_tmf.cpp:69]: If it doesn't cost much to be accurate, why leave it vague?
I may prefer a bit to update the code to be consistent with the rules `1.`, `2.` you defined in the comments.

[the remark this answers is not in the document]
Comment (simonlingoogle) [reply src/core/backbone_router/backbone_tmf.cpp:69]: revised accordingly.

File: src/core/backbone_router/bbr_manager.cpp [hunk 13]
```diff
@@ -86,6 +89,33 @@ void Manager::HandleNotifierEvents(Events aEvents)
                 mTimer.Start(kTimerInterval);
             }
         }
+
+        if (Get<BackboneRouter::Local>().GetState() == OT_BACKBONE_ROUTER_STATE_PRIMARY)
```
[the remark this answers is not in the document]
Comment (simonlingoogle) [reply src/core/backbone_router/bbr_manager.cpp]: revised accordingly.

File: src/core/backbone_router/backbone_tmf.hpp [hunk 14]
```diff
@@ -0,0 +1,91 @@
...
+     * @retval OT_ERROR_ALREADY Already started.
+     *
+     */
+    otError Start(void);
+
+    /**
+     * This method returns whether Thread Management Framework Addressing Rules are met.
+     *
+     * @retval TRUE   if Thread Management Framework Addressing Rules are met.
+     * @retval FALSE  if Thread Management Framework Addressing Rules are not met.
```
[the remark this answers is not in the document]
Comment (simonlingoogle) [reply src/core/backbone_router/backbone_tmf.hpp]: revised accordingly.

**Review by bukepo — APPROVED** [review]
LGTM:+1:

File: src/core/net/udp6.cpp [hunk 15]
```diff
@@ -156,7 +172,10 @@ otError Udp::RemoveReceiver(Receiver &aReceiver)
 
 otError Udp::Open(SocketHandle &aSocket, otUdpReceive aHandler, void *aContext)
 {
-    otError error = OT_ERROR_NONE;
+    otError       error = OT_ERROR_NONE;
+    SocketHandle *prev;
+
+    VerifyOrExit(mSockets.Find(aSocket, prev) == OT_ERROR_NOT_FOUND, error = OT_ERROR_ALREADY);
```
Comment (simonlingoogle) [inline src/core/net/udp6.cpp:178]: Turn out `Udp::Open` is not **reentrant**. 
Currently `BackboneTmf` agent would call `Coap::Start` when BBR primary state changes, which in turn calls `Udp::Open`.
So I make `Udp::Open` **reentrant** as the code shows. 
@librasungirl  @abtink @bukepo

File: src/core/coap/coap.cpp [hunk 16]
```diff
@@ -1020,9 +1020,11 @@ Coap::Coap(Instance &aInstance)
 
 otError Coap::Start(uint16_t aPort, otNetifIdentifier aNetifIdentifier)
 {
-    otError error;
+    otError error        = OT_ERROR_NONE;
     bool    socketOpened = false;
 
+    VerifyOrExit(!mSocket.IsBound(), OT_NOOP);
```
Comment (librasungirl) [inline src/core/coap/coap.cpp:1026]: According to the available return values in the header file,  `OT_ERROR_ALREADY` might be more appropriate here?

https://github.com/openthread/openthread/blob/25ae9b073955a4fa00d8b0b2ed368c1ac77fec8f/src/core/coap/coap.hpp#L607-L616

**Review by librasungirl — APPROVED** [review]
👍  LGTM with one minor inline comment

File: src/posix/platform/system.cpp [hunk 17]
```diff
@@ -63,6 +63,8 @@ otInstance *otSysInit(otPlatformConfig *aPlatformConfig)
     instance = otInstanceInitSingle();
     assert(instance != nullptr);
 
+    platformBackboneInit(instance, aPlatformConfig->mBackboneInterfaceName);
```
Comment (librasungirl) [inline src/posix/platform/system.cpp:66]: add `OPENTHREAD_FTD && OPENTHREAD_CONFIG_BACKBONE_ROUTER_ENABLE` condition?

File: src/posix/platform/backbone.cpp [hunk 18]
```diff
@@ -0,0 +1,54 @@
...
+    VerifyOrExit(aInterfaceName != nullptr, OT_NOOP);
+
+    VerifyOrDie(strnlen(aInterfaceName, IFNAMSIZ) <= IFNAMSIZ - 1, OT_EXIT_INVALID_ARGUMENTS);
+    strcpy(gBackboneNetifName, aInterfaceName);
+
+    gBackboneNetifIndex = if_nametoindex(gBackboneNetifName);
+    VerifyOrDie(gBackboneNetifIndex > 0, OT_EXIT_FAILURE);
+exit:
+    return;
+}
```
Comment (librasungirl) [inline src/posix/platform/backbone.cpp:54]: add `OPENTHREAD_FTD && OPENTHREAD_CONFIG_BACKBONE_ROUTER_ENABLE` condition?

File: src/posix/platform/platform-posix.h [hunk 19]
```diff
@@ -395,6 +395,15 @@ enum SocketBlockOption
...
 int SocketWithCloseExec(int aDomain, int aType, int aProtocol, SocketBlockOption aBlockOption);
 
+/**
+ * This function initializes platform Backbone network.
+ *
+ * @param[in]   aInstance       A pointer to the OpenThread instance.
+ * @param[in]   aInterfaceName  A pointer to Thread network interface name.
+ *
+ */
+void platformBackboneInit(otInstance *aInstance, const char *aInterfaceName);
```
Comment (librasungirl) [inline src/posix/platform/platform-posix.h:405]: add `OPENTHREAD_FTD && OPENTHREAD_CONFIG_BACKBONE_ROUTER_ENABLE` condition?

File: src/core/coap/coap.cpp (hunk 16 above)
Comment (simonlingoogle) [reply src/core/coap/coap.cpp:1026]: Agree that it's better for this PR to return `OT_ERROR_ALREADY`. 

But I think in general we should avoid using `OT_ERROR_ALREADY`. We can refactor it in a separate PR.

File: src/posix/platform/system.cpp (hunk 17 above)
Comment (simonlingoogle) [reply src/posix/platform/system.cpp:66]: It's actually not a must. 
- It's not a strong requirement of posix platform code to reduce the code size, so we can have this code running even when it's not a Backbone Router. 
- Further more, it's possible (and perfectly fine) that other COAP objects (other than Backbone TMF agent) might need to bind on the Backbone interface.

I don't have a strong opinion for it, but I also think it's fine to keep the code available on non Backbone Routers. 

Thoughts? @librasungirl

File: src/core/backbone_router/backbone_tmf.hpp [hunk 20]
```diff
@@ -0,0 +1,91 @@
...
+    otError Start(void);
+
+    /**
+     * This method returns whether Thread Management Framework Addressing Rules are met.
+     *
+     * @retval true   Thread Management Framework Addressing Rules are met.
+     * @retval false  Thread Management Framework Addressing Rules are not met.
+     *
+     */
+    bool IsBackboneTmfMessage(const Ip6::MessageInfo &aMessageInfo);
```
Comment (jwhui) [inline src/core/backbone_router/backbone_tmf.hpp:82]: Declare as const?

File: src/posix/platform/system.cpp (hunk 17 above)
Comment (librasungirl) [reply src/posix/platform/system.cpp:66]: I may prefer to add the macro condition: 
- follow the same style as the others, like `platformNetifProcess()` has `OPENTHREAD_CONFIG_PLATFORM_NETIF_ENABLE` wrapper
- As similar rule in another PR https://github.com/openthread/openthread/pull/5346#discussion_r475161377 agreed -better to only add required functionalities unless we have clear use case for the `possible (and perfectly fine) that other COAP objects (other than Backbone TMF agent) might need to bind on the Backbone interface`, though we don't concern code size on posix so much as in embedded platform.

File: src/core/backbone_router/backbone_tmf.hpp [hunk 21]
```diff
@@ -0,0 +1,91 @@
...
+     * This method starts the TMF agent.
+     *
+     * @retval OT_ERROR_NONE    Successfully started the CoAP service.
+     * @retval OT_ERROR_ALREADY Already started.
+     *
+     */
+    otError Start(void);
+
+    /**
+     * This method returns whether Thread Management Framework Addressing Rules are met.
```
Comment (Irving-cl) [inline src/core/backbone_router/backbone_tmf.hpp:76]:
```suggestion
     * This method returns whether @p aMessageInfo meets Thread Management Framework Addressing Rules.
```
What about saying like this?

File: src/core/common/instance.hpp [hunk 22]
```diff
@@ -763,10 +763,16 @@ template <> inline BackboneRouter::MulticastListenersTable &Instance::Get(void)
...
+
 template <> inline BackboneRouter::NdProxyTable &Instance::Get(void)
 {
     return mThreadNetif.mBackboneRouterManager.GetNdProxyTable();
 }
+
+template <> inline BackboneRouter::BackboneTmfAgent &Instance::Get(void)
+{
+    return mThreadNetif.mBackboneRouterManager.GetBackboneTmfAgent();
+}
```
Comment (Irving-cl) [inline src/core/common/instance.hpp:775]: This seems not used anywhere.

File: src/core/net/udp6.cpp [hunk 23]
```diff
@@ -99,13 +99,26 @@ otError Udp::Socket::Bind(uint16_t aPort)
...
-}
+    OT_UNUSED_VARIABLE(aNetifIdentifier);
+
+    otError error = OT_ERROR_NONE;
+
+#if OPENTHREAD_CONFIG_PLATFORM_UDP_ENABLE
+    SuccessOrExit(error = otPlatUdpBindToNetif(this, aNetifIdentifier));
+#endif
+
+#if OPENTHREAD_FTD && OPENTHREAD_CONFIG_BACKBONE_ROUTER_ENABLE
```
Comment (Irving-cl) [inline src/core/net/udp6.cpp:112]: Change to `#if OPENTHREAD_CONFIG_BACKBONE_ROUTER_ENABLE` and add config check to see whether this is enabled when `OPENTHREAD_FTD` is not enabled in `config-check.h`?

Example: https://github.com/openthread/openthread/blob/master/src/core/config/openthread-core-config-check.h#L526

File: src/core/net/udp6.cpp [hunk 24]
```diff
@@ -296,14 +341,34 @@ otError Udp::SendTo(SocketHandle &aSocket, Message &aMessage, const MessageInfo
...
 {
-    SuccessOrExit(mSockets.Remove(aSocket));
+    SocketHandle *prev;
+
+    SuccessOrExit(mSockets.Find(aSocket, prev));
+
+    mSockets.PopAfter(prev);
     aSocket.SetNext(nullptr);
 
+#if OPENTHREAD_FTD && OPENTHREAD_CONFIG_BACKBONE_ROUTER_ENABLE
```
Comment (Irving-cl) [inline src/core/net/udp6.cpp:365]: The same as mentioned above.

File: src/core/net/udp6.cpp [hunk 25]
```diff
@@ -397,7 +462,27 @@ void Udp::HandlePayload(Message &aMessage, MessageInfo &aMessageInfo)
     SocketHandle *socket;
     SocketHandle *prev;
 
+#if OPENTHREAD_FTD && OPENTHREAD_CONFIG_BACKBONE_ROUTER_ENABLE
```
Comment (Irving-cl) [inline src/core/net/udp6.cpp:465]: The same as mentioned above.

File: src/core/common/instance.hpp (hunk 22 above)
Comment (simonlingoogle) [reply src/core/common/instance.hpp:775]: It will be used soon in #5388

File: src/core/net/udp6.cpp (hunk 23 above)
Comment (simonlingoogle) [reply src/core/net/udp6.cpp:112]: This is indeed a good idea. But I prefer to do it in a separate PR because there are so many places that are using `#ifdef OPENTHREAD_FTD && OPENTHREAD_CONFIG_BACKBONE_ROUTER_ENABLE` and they should all be fixed (but unrelated to this PR).

Comment (Irving-cl) [reply src/core/net/udp6.cpp:112]: SGTM.

**Review by Irving-cl — APPROVED** [review]
LGTM👍

**Review by jwhui — APPROVED** [review]
